subsys/bootloader/fw_metadata/fw_metadata.c

```diff
@@ -5,15 +5,68 @@
  */
 
 #include "fw_metadata.h"
+#include <linker/sections.h>
+#include <errno.h>
+#include <string.h>
+
 
 extern const u32_t _image_rom_start;
 extern const u32_t _flash_used;
+extern const struct fw_firmware_info _firmware_info_start[];
+extern const struct fw_abi_info _extabis_start[];
+extern const struct fw_abi_info _extabis_end[];
+
+/* Go to next extabi, based on the length in the header. */
+#define EXTABIS_NEXT(extabi) ((const struct fw_abi_info *)((u32_t)extabi + extabi->abi_len))
+
+int abi_getter(u32_t id, u32_t index, const struct fw_abi_info **abi)
+{
+	if (!abi) {
+		return -EFAULT;
+	}
 
-const struct fw_firmware_info m_firmware_info
-Z_GENERIC_SECTION(.firmware_info)
-__attribute__((used)) = {
+	bool id_found = false;
+
+	for (const struct fw_abi_info *extabi = _extabis_start;
+	     extabi < _extabis_end; extabi = EXTABIS_NEXT(extabi))
+	{
+		if (extabi->abi_id == id) {
+			id_found = true;
+			if (index-- == 0) {
+				*abi = extabi;
+				return 0;
+			}
+		}
+	}
+	return id_found ? -EBADF : -ENOENT;
+}
+
+const struct fw_abi_getter_info ABI_GETTER_OUT_NAME = {
+	.magic = {ABI_GETTER_INFO_MAGIC},
+	.abi_getter = abi_getter,
+};
+__noinit struct fw_abi_getter_info ABI_GETTER_IN_NAME;
+
+
+Z_GENERIC_SECTION(.firmware_info) __attribute__((used))
+const struct fw_firmware_info m_firmware_info =
+{
 	.magic = {FIRMWARE_INFO_MAGIC},
 	.firmware_size = (u32_t)&_flash_used,
 	.firmware_version = CONFIG_SB_FIRMWARE_VERSION,
-	.firmware_address = (u32_t)&_image_rom_start
+	.firmware_address = (u32_t)&_image_rom_start,
+	.abi_in = &ABI_GETTER_IN_NAME,
+	.abi_out = &ABI_GETTER_OUT_NAME,
 };
+
+void abi_provide(u32_t address)
+{
+	const struct fw_firmware_info *our_fw_info = _firmware_info_start;
```

Comment:
is _firmware_info_start the same as &m_firmware_info ?

Reply:
If so, we could drop the custom section, or?

Reply:
No. There are actually two custom sections at play: `.firmware_info` has been around for a while and needs to be placed at a specific offset inside the image (currently 0x800). `extabis` has been added for this patch and is meant as an aggregator for exposed abis. Any module can declare an ABI inside the extabis sections, and the list will be provided to the next image.

Reply:
Ah, my mistake, I assumed Z_GENERIC_SECTION was declaring a section somehow, and also adding the .firmware_info attribute to m_firmware_info.

Reply:
That would've been nice :)